cranelift/codegen/src/isa/riscv64/inst/emit.rs

```diff
@@ -556,6 +556,113 @@ impl Inst {
+            // where the immediate is scaled to represent multiples of 16 in the range (-512,496). c.addi16sp is used
+            // to adjust the stack pointer in procedure prologues and epilogues. It expands into addi x2, x2, nzimm. c.addi16sp
+            // is only valid when nzimm≠0; the code point with nzimm=0 is reserved.
+            Inst::AluRRImm12 {
+                alu_op: AluOPRRI::Addi,
+                rd,
+                rs,
+                imm12,
+            } if has_zca
+                && rd.to_reg() == rs
+                && rs == stack_reg()
+                && imm12.as_i16() != 0
+                && (imm12.as_i16() % 16) == 0
+                && Imm6::maybe_from_i16(imm12.as_i16() / 16).is_some() =>
+            {
+                let imm6 = Imm6::maybe_from_i16(imm12.as_i16() / 16).unwrap();
+                sink.put2(encode_c_addi16sp(imm6));
+            }
+
+            // c.addi4spn
+            //
+            // c.addi4spn is a CIW-format instruction that adds a zero-extended non-zero
+            // immediate, scaled by 4, to the stack pointer, x2, and writes the result to
+            // rd. This instruction is used to generate pointers to stack-allocated variables
+            // and expands to addi rd, x2, nzuimm. c.addi4spn is only valid when nzuimm≠0;
+            // the code points with nzuimm=0 are reserved.
+            Inst::AluRRImm12 {
+                alu_op: AluOPRRI::Addi,
+                rd,
+                rs,
+                imm12,
+            } if has_zca
+                && reg_is_compressible(rd.to_reg())
+                && rs == stack_reg()
+                && imm12.as_i16() != 0
+                && (imm12.as_i16() % 4) == 0
+                && u8::try_from(imm12.as_i16() / 4).is_ok() =>
+            {
+                let imm = u8::try_from(imm12.as_i16() / 4).unwrap();
+                sink.put2(encode_ciw_type(CiwOp::CAddi4spn, rd, imm));
+            }
+
+            // c.addi
+            Inst::AluRRImm12 {
+                alu_op: AluOPRRI::Addi,
+                rd,
+                rs,
+                imm12,
+            } if has_zca
+                && rd.to_reg() == rs
+                && rs != zero_reg()
+                && imm12.as_i16() != 0
+                && Imm6::maybe_from_imm12(imm12).is_some() =>
+            {
+                let imm6 = Imm6::maybe_from_imm12(imm12).unwrap();
+                sink.put2(encode_ci_type(CiOp::CAddi, rd, imm6));
+            }
+
+            // c.addi / c.sext.w
+            Inst::AluRRImm12 {
+                alu_op: AluOPRRI::Addiw,
+                rd,
+                rs,
+                imm12,
+            } if has_zca
+                && rd.to_reg() == rs
+                && rs != zero_reg()
+                && Imm6::maybe_from_imm12(imm12).is_some() =>
+            {
+                let imm6 = Imm6::maybe_from_imm12(imm12).unwrap();
```

Comment:
To avoid the double-construction logic for `Imm6::maybe_from_imm12` could this do:

```rust
let imm6 = match Imm6::maybe_from_imm12(imm12) {
    Some(imm6) => imm6,
    None => return false,
};
```

or alternatively if this function returned `Option<()>` instead of `bool` it could use `?`. Or alternatively I think this could use `let Some(imm6) = ... else { return false; };`, but I haven't use `let`/`else` much myself.

Reply:
Here I think it's ok  to do that, but I'd like to keep the separate check for `c.addi16sp` and `c.addi4spn` so that we can try to match all 3 instructions. Otherwise we could enter into one branch and never emit the instruction.